src/circuit_builder.rs

```diff
@@ -0,0 +1,166 @@
+use std::{
+    cmp::max,
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use crate::circuit::{Gate, GateAddr, GateOp, GeneralCircuit, Layer};
+
+#[derive(Debug, Clone)]
+struct Builder {
+    // contains a vec of all Layers
+    layers: Vec<Vec<Node>>,
+}
+
+#[derive(Debug, Clone)]
+struct Node {
+    // This is the operation of the gate
+    // TODO: make this a reference to an operation
+    op: Option<GateOp>,
+    // Used for checking if the node has been processed
+    is_processed: bool,
+    // The left child of the node
+    left_child: Option<GateAddr>,
+    // The right side of the node
+    right_child: Option<GateAddr>,
+    // The layer index and gate index of the node
+    node_addr: Option<GateAddr>,
+}
+
+impl Node {
+    pub fn new(
+        op: Option<GateOp>,
+        left_child: Option<GateAddr>,
+        right_child: Option<GateAddr>,
+        node_addr: Option<GateAddr>,
+    ) -> Node {
+        Self {
+            op,
+            is_processed: false,
+            left_child,
+            right_child,
+            node_addr,
+        }
+    }
+
+    fn is_input(&self) -> bool {
+        self.left_child.is_none() && self.right_child.is_none() && self.op.is_none()
+    }
+
+    fn to_gate(&self) -> Gate {
```

Comment:
This has the potential to fail. Do we have guarantees that unwrap will not break

Reply:
okay nevermind, just saw that you skip the input layer so .to_gate() should never be called on input elements.